sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -1726,8 +1726,7 @@ object CodeGenerator extends Logging {
     if (nullable) {
       // Can't call setNullAt on DecimalType/CalendarIntervalType, because we need to keep the
       // offset
-      if (!isVectorized && (dataType.isInstanceOf[DecimalType] ||
-        dataType.isInstanceOf[CalendarIntervalType])) {
+      if (!isVectorized && UnsafeRowUtils.avoidSetNullAt(dataType)) {
```

Comment:
This will slightly change the conditions under which `CodeGenerator` will avoid generating `row.setNullAt` for decimals. Before, `CodeGenerator` always avoided calling `row.setNullAt` for decimals. Now it will only do so if the decimals precision is greater than `Decimal.MAX_LONG_DIGITS` (which is the only case where it matters, AFAIK).